### PR TITLE
[MAINTENANCE] Add DatasourceStore:delete_batch_config

### PR DESCRIPTION
**Pull request by tyler-hoffman**
**Main work**
Implements `DatasourceStore::delete_batch_config`.

**Cleanup along the way**
* Adds some fixtures to DRY up the new tests along with those around `add_batch_config`
* Moves the wiring of `batch_config._data_asset` from config.py to the Datasource class itself. This ensures that batch configs are wired to data assets in both cases:
  * Loading config from yaml/json/dict
  * Loading from DatasourceStore

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/store/datasource_store.py [hunk 1]
```diff
@@ -215,6 +215,16 @@ def _build_key_from_config(  # type: ignore[override]
         )
         return self.store_backend.build_key(name=datasource_config.name, id=id_)
 
+    def get_fluent_datasource_by_name(self, name: str) -> FluentDatasource:
```
Comment (tyler-hoffman) [inline great_expectations/data_context/store/datasource_store.py:218]: I got sick of writing `isinstance(datasource, Datasource)` everywhere, so this at least removes some of the need for that.

File: tests/data_context/store/test_datasource_store.py [hunk 2]
```diff
@@ -48,6 +48,42 @@ def fake_datasource_name() -> str:
     return "my_first_datasource"
 
 
+@pytest.fixture
```
Comment (tyler-hoffman) [inline tests/data_context/store/test_datasource_store.py:51]: I've seen us make fixtures in other places, including this file, so I stuck with that pattern. Probably cleaner to just make top-level constants, unless there's something I'm missing. I'll leave it to the reviewer to make the call.

File: great_expectations/datasource/fluent/pandas_datasource.py [hunk 3]
```diff
@@ -85,6 +85,7 @@ class PandasDatasourceError(Exception):
 
 class _PandasDataAsset(DataAsset):
     _EXCLUDE_FROM_READER_OPTIONS: ClassVar[Set[str]] = {
+        "batch_configs",
```
Comment (tyler-hoffman) [inline great_expectations/datasource/fluent/pandas_datasource.py:88]: Without this, we end up passing `batch_configs` to read_* methods, which are not expecting that key.

File: great_expectations/data_context/store/datasource_store.py [hunk 4]
```diff
@@ -354,3 +363,28 @@ def add_batch_config(self, batch_config: BatchConfig) -> BatchConfig:
 
         updated_batch_config._data_asset = updated_asset
         return updated_batch_config
+
+    def delete_batch_config(self, batch_config: BatchConfig) -> None:
```
Comment (joshua-stauffer) [inline great_expectations/data_context/store/datasource_store.py:367]: what's our convention around what these methods should return? i've been returning the deleted object in `ExpectationsStore` - don't have a strong opinion either way right now, but let's coordinate for consistency.

File: tests/data_context/store/test_datasource_store.py [hunk 5]
```diff
@@ -227,29 +264,71 @@ def test_datasource_store__add_batch_config__success(
...
+        empty_asset_name
+    )
+
+    # Act + Assert
+    new_batch_config = BatchConfig(name=batch_config_name)
+    new_batch_config._data_asset = asset
+
+    with pytest.raises(ValueError) as e:
+        store.delete_batch_config(new_batch_config)
+    assert "does not exist" in str(e.value)
```
Comment (joshua-stauffer) [inline tests/data_context/store/test_datasource_store.py:329]:
```suggestion
    with pytest.raises(ValueError, match="does not exist"):
        store.delete_batch_config(new_batch_config)

```

**Review by joshua-stauffer — APPROVED** [review]
this looks great, thanks!

File: great_expectations/data_context/store/datasource_store.py (hunk 4 above)
Comment (tyler-hoffman) [reply great_expectations/data_context/store/datasource_store.py:367]: I don't have strong opinions either, but my thinking was that add returns an object representing what is now persisted, and for parity, delete_* should return what's persisted: nothing.

**Pull request merged**